aten/src/ATen/native/cuda/Dropout.cu

```diff
@@ -80,10 +154,50 @@ fused_dropout_kernel(cuda::detail::TensorInfo<scalar_t, IndexType> a,
 
 template<typename scalar_t, typename accscalar_t>
 void masked_scale_kernel(at::Tensor& ret, const at::Tensor src, const at::Tensor mask, accscalar_t scale){
-   at::cuda::CUDA_tensor_apply3<scalar_t, scalar_t, uint8_t>(ret, src, mask, [scale]__device__(scalar_t& ret_val, const scalar_t& src_val, const uint8_t mask_val){
-       ret_val = (float)mask_val * src_val * scale;
-  });
+   auto iter = at::TensorIterator();
+   iter.add_output(ret);
+   iter.add_input(src);
+   iter.add_input(mask);
+   iter.dont_compute_common_dtype();
+
+   iter.build();
+
+   at::native::gpu_kernel(
+       iter,
+       [=]GPU_LAMBDA(const scalar_t src_val, const uint8_t mask_val) -> scalar_t {
+          return (float)mask_val * src_val * scale;
+       });
+}
+
+int get_alignment(uint64_t addr) {
+  if (addr % 16 == 0) return 16;
+  if (addr %  8 == 0) return  8;
+  if (addr %  4 == 0) return  4;
+  if (addr %  2 == 0) return  2;
+  return 1;
 }
+
+template <typename scalar_t>
+int get_vector_size(at::Tensor self, at::Tensor ret, at::Tensor mask) {
+  int vec_size = 4;
+  // get the vector size
+  if (!self.is_contiguous() || !ret.is_contiguous() || !mask.is_contiguous()) {
+    vec_size = 1;
+  } else {
+    vec_size = std::min(vec_size, (get_alignment((uint64_t)self.data_ptr()) / (int)sizeof(scalar_t)));
+    vec_size = std::min(vec_size, get_alignment((uint64_t)ret.data_ptr()) / (int)sizeof(scalar_t));
+    vec_size = std::min(vec_size, get_alignment((uint64_t)mask.data_ptr()) / (int)sizeof(uint8_t));
+  }
+
+  // check that we'd have no remainders - prefer a smaller vector size with no remainders over a larger vector and remainder.
```

Comment:
Hmm... Compared to reduce the vec size, would it be better to handle the reminder inside the kernel? In a real use case, for example, ResNet-50, how frequently would this cause a drop of vec size?

Reply:
I'm not sure to be honest -- I went this route for code simplicity, if we feel it would be better in a single (more complex) kernel I'm ok with that. In RN50 and such you'll always be in the VEC=4 regime purely due to the channels being multiples of 4, so `N * C * H * W` should also be a multiple of 4

Reply:
I just checked BERT also at https://github.com/huggingface/transformers/blob/master/src/transformers/configuration_bert.py
It should also be a multiple of 4. The leave this as is, and see what Facebook think about it.